### PR TITLE
Add GLV scalar multiplication for both Pasta curves

### PR DESCRIPTION
**Pull request by LukasKorba**
## GLV scalar multiplication for both Pasta curves

Follow-up to the review discussion on zcash/orchard#530: hosting the endomorphism-based
multiplication machinery at the curve layer, where `CurveExt::endo()` and the `ZETA`
constants already live.

### What this adds

A new `glv` module (additive — the native constant-time `Mul` implementations are
untouched, with `_glv`-suffixed naming per review guidance):

- **GLV decomposition**: `k = k1 + k2·λ (mod r)` with |k1|, |k2| ≤ 2^127, via a short
  lattice basis and 2^384 Babai rounding, followed by width-4 wNAF recoding.
- **`glv::Table`**: the odd multiples {1,3,5,7}·P and {1,3,5,7}·φ(P) (512 bytes).
  `Table::batch` builds many tables with **one shared batch normalization** — a single
  field inversion for the whole batch.
- **`glv::Decomposed`**: a per-scalar handle hoisting decomposition + recoding out of
  loops that multiply one scalar against many tables.
- **`MulGlv::mul_glv`**: one-shot convenience.

The three cost centers are exposed composably because the downstream win (batched trial
decryption in orchard) depends on amortizing them independently: tables per ephemeral
key (batched), decomposition per viewing key (hoisted), ladder per pair.

### Toolchain

Based directly on `main` — this PR carries no dependency on #90: edition 2021,
`rust-version = "1.63"`, lockfile v3. The CI matrix was reproduced locally under the
repository's pinned 1.63.0 toolchain: tests `--all-features` and `--no-default-features`,
`clippy --all-features --all-targets -- -D warnings`, `cargo fmt -- --check`,
`cargo doc --all-features --document-private-items`, and `--no-default-features` builds
for `thumbv6m-none-eabi` / `wasm32-unknown-unknown` / `wasm32-wasi`.

`proptest` (new dev-dependency) is declared with minimal features (`std` only) and held
in `Cargo.lock` at 1.2.0 — the newest release that builds on the MSRV toolchain.

### Correctness

- The lattice constants are **re-verified in-crate** against each curve's own λ
  (= `Scalar::ZETA`) using field arithmetic only (`constants` tests) — consistent with
  the existing `test_endo_consistency` pinning of φ ↔ λ.
- `decompose` tests prove `k1 + k2·λ ≡ k` with both halves ≤ 2^127 over 1000 full-width
  scalars plus edge cases (0, ±1, ±λ), per curve. Wrong constants cannot pass.
- Table multiplication, one-shot `mul_glv`, batched-vs-solo table builds, and
  hoisted-vs-fresh decomposition are each asserted equal to the native operator /
  each other, per curve (26 tests total: KATs + edge matrices + 8 proptest properties).
- Pallas constants match the reference implementation on zcash/orchard#530. Vesta
  constants were derived by the same lattice reduction, validated by first reproducing
  the Pallas set bit-for-bit from `Fq::ZETA`.

### Property-based tests

Per review request, a `proptest` suite (dev-dependency) drives the FULL public path on
both curves — scalars drawn as four uniform u64 limbs widened through
`from_uniform_bytes`, so the whole field is reachable without modular bias:

- ∀ P≠O, k: `P.mul_glv(&k) == P * k` (the headline property);
- ∀ k: the GLV split reconstructs `k` with both halves ≤ 2^127;
- ∀ point sets: `Table::batch` ≡ solo `Table::new`, pointwise, including products;
- ∀ k reused across points: hoisted `Decomposed` ≡ fresh decomposition.

Plus an explicit edge-case matrix through `mul_glv`: k ∈ {0, 1, −1, 2, λ, −λ, λ+1,
2^64−1, 2^127−1, 2^127, 2^127+1} against the generator, a random point, **and the
identity** — `mul_glv` is total: `k·O = O` (early return; tables keep their documented
non-identity contract).

### Security

This path is **variable-time in the scalar** (documented on the module): intended for
public/per-session scalars (batched trial decryption, verification-style workloads),
not long-term secrets. The constant-time `Mul` is unchanged and remains the default.

### Benchmarks (Apple M4, rustc 1.63.0 — the pinned toolchain; `cargo bench --bench glv`)

| Path | Pallas | Vesta | vs native |
|---|---|---|---|
| native `Mul` (constant-time) | 109.9 µs | 112.8 µs | 1× |
| `mul_glv` one-shot | 39.4 µs | 39.3 µs | **2.8×** |
| reused `Table` + `Decomposed` | 30.2 µs | 29.6 µs | **3.6–3.8×** |
| `Table::new` (solo) | 8.8 µs | 8.7 µs | — |
| `Table::batch` ×50, per point | 2.07 µs | 2.07 µs | 4.3× cheaper than solo |

(The larger relative win vs orchard#530's −38–40% is expected: the baseline here is the
crate's constant-time double-and-add, not orchard's variable-time wNAF preparation.)

### Relationship to zcash/orchard#530

With this merged and released, #530 slims to protocol integration only (prepared-EPK
variants + the `zcash_note_encryption` batch hooks) consuming this API, and Vesta gets
the same speedup for free.

### CODE REVIEW

**Comment by LukasKorba** [comment]
Re-based onto `main` — this PR is now self-contained and no longer depends on #90, per
the decision to keep the current toolchain. Same single commit, same code; the only
source delta vs the previous head is one import in `benches/glv.rs` reordered back to
the 2021 rustfmt style.

Toolchain contract as on `main`: edition 2021, `rust-version = "1.63"`, lockfile v3
(the previous head's lock had been bumped to v4, which the pinned cargo can't read —
regenerated). Verified locally under the repository's pinned 1.63.0 toolchain:

- `cargo test --release --all-features` (74 + doc-test) and `--no-default-features` (22) — all passing
- `cargo clippy --all-features --all-targets -- -D warnings` — clean
- `cargo fmt -- --check` — clean
- `cargo doc --all-features --document-private-items` — clean
- `--no-default-features` builds: `thumbv6m-none-eabi`, `wasm32-unknown-unknown`, `wasm32-wasi`

Dev-dependency note: `proptest` is declared with minimal features (`std` only) and held
in `Cargo.lock` at 1.2.0 — the newest release that builds on the MSRV toolchain.

Benchmarks re-measured under 1.63.0 (absolute numbers shift ~+14% with the older
codegen; every ratio is unchanged: 2.8× one-shot, 3.6–3.8× amortized, 4.3× cheaper
batched table builds) — the PR description's table is updated.

Marking ready for review: the draft status only reflected the former #90 stacking.

File: src/glv.rs [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+//! Both Pasta curves are equipped with a cube-root endomorphism
+//! φ(x, y) = (ζ·x, y) (exposed as [`CurveExt::endo`]), for which
+//! φ(P) = λ·P with λ = `Scalar::ZETA`. This module uses that structure to
+//! split a full-width scalar multiplication `k·P` into two half-width
+//! multiplications evaluated against a shared table of odd multiples of `P`
+//! and φ(P) — the classic GLV speedup described in the Halo paper.
+//!
+//! The API is additive: the `_glv` naming distinguishes this variable-time
+//! path from the constant-time native `Mul` implementations, which are
+//! unchanged.
```
Comment (ebfull) [inline src/glv.rs:12]: the current implementation in `Mul` is not guaranteed constant time though?

File: src/glv.rs [hunk 2]
```diff
@@ -0,0 +1,711 @@
+//! GLV (Gallant–Lambert–Vanstone) scalar multiplication for the Pasta curves.
+//!
+//! Both Pasta curves are equipped with a cube-root endomorphism
+//! φ(x, y) = (ζ·x, y) (exposed as [`CurveExt::endo`]), for which
+//! φ(P) = λ·P with λ = `Scalar::ZETA`. This module uses that structure to
+//! split a full-width scalar multiplication `k·P` into two half-width
+//! multiplications evaluated against a shared table of odd multiples of `P`
+//! and φ(P) — the classic GLV speedup described in the Halo paper.
```
Comment (ebfull) [inline src/glv.rs:8]: We shouldn't be adding unicode greek symbols in rustdoc comments that get rendered in rustdoc. We have katex rendering, so these should be using math formulas.

File: src/glv.rs [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+//! The API is additive: the `_glv` naming distinguishes this variable-time
+//! path from the constant-time native `Mul` implementations, which are
+//! unchanged.
+//!
+//! # Security
+//!
+//! **This path is variable-time with respect to the scalar** (GLV
+//! decomposition plus wNAF recoding). It is intended for public or
+//! per-session scalars — batched trial decryption, verification-style
+//! workloads — and must not be used with long-term secret scalars.
```
Comment (ebfull) [inline src/glv.rs:19]: We don't add disclaimers like this elsewhere in the project; it's assumed that unless something is described as constant time, it isn't, but this is probably harmless.

File: src/glv.rs [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+}
+
+/// Per-curve GLV constants: a short basis for the lattice
+/// {(a, b) : a + b·λ ≡ 0 (mod r)} and the Babai rounding coefficients
+/// derived from it.
+///
+/// The `constants_verify_*` tests re-verify every constant against the
+/// curve's own λ (= `Scalar::ZETA`) using field arithmetic only, and the
+/// `decompose_reconstructs_*` tests prove the full pipeline algebraically:
+/// wrong constants cannot reconstruct `k`.
```
Comment (ebfull) [inline src/glv.rs:58]: These tests should not leak into the documentation for the trait. They can be mentioned on the specific implementations.

File: src/glv.rs [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+    /// Magnitude of v1's (negative) second component.
+    const V1B_NEG: u128;
+    /// v2 = (`V2A`, `V2B`): second short lattice vector.
+    const V2A: u128;
+    /// v2's (positive) second component.
+    const V2B: u128;
+    /// Babai coefficient `round(2^384·V2B / r)`, little-endian limbs.
+    const G1: [u64; 5];
+    /// Babai coefficient `round(2^384·V1B_NEG / r)`, little-endian limbs.
+    const G2: [u64; 5];
```
Comment (ebfull) [inline src/glv.rs:74]: The constants in this trait aren't explained well; what is `r`? We use `r` in this repository to already refer to things related to the Montgomery reduction constant, which a reader might confuse with it.

File: src/glv.rs [hunk 6]
```diff
@@ -0,0 +1,711 @@
+//! GLV (Gallant–Lambert–Vanstone) scalar multiplication for the Pasta curves.
```
Comment (ebfull) [inline src/glv.rs:1]: We should link to a paper or resource for learning about GLV to put the remainder of this PR into context.

**Review by ebfull — COMMENTED** [review]
In #87 we added a different kind of optimization that also could be provided as a purely additive change to the API, but we did so under a feature flag so the optimizations had time to be adopted before the API was settled. This was conservative.

I would advocate either for:

1. we feature flag the new glv APIs in this PR (for consistency)
2. we remove the feature flag for #87 before the next release (for consistency)

One of the reasons you might want to be conservative here is that GLV is only one method for doing the kinds of operations desired. There could be more efficient operations that don't use GLV or rather use an extension of it that wouldn't be accurately described by GLV. Since the example is handy, #87 doesn't expose any of the inner details of its optimization in its API.

File: src/glv.rs [hunk 7]
```diff
@@ -0,0 +1,711 @@
...
+        }
+        prod[i + 4] = prod[i + 4].wrapping_add(carry as u64);
+    }
+    // Bits >= 384 live in limbs 6..; round on bit 383 (top bit of limb 5).
+    let round = prod[5] >> 63;
+    (u128::from(prod[6]) | (u128::from(prod[7]) << 64)).wrapping_add(u128::from(round))
+}
+
+/// 256-bit product of two `u128`s, as little-endian limbs.
+fn mul_u128(a: u128, b: u128) -> [u64; 4] {
```
Comment (ebfull) [inline src/glv.rs:138]: This implementation is variable time, but not documented or named as such.

File: src/glv.rs [hunk 8]
```diff
@@ -0,0 +1,711 @@
...
+        borrow = u64::from(b1) + u64::from(b2);
+    }
+    out
+}
+
+/// Interprets a 256-bit two's-complement value with |x| < 2^128 as
+/// (is_negative, |x|).
+fn signed_halves(x: [u64; 4]) -> (bool, u128) {
+    if x[3] >> 63 == 0 {
+        debug_assert!(x[2] == 0 && x[3] == 0, "positive half exceeds 128 bits");
```
Comment (ebfull) [inline src/glv.rs:176]: This assertion can be triggered by the caller, but only in debug mode. That's two problems.

File: src/glv.rs [hunk 9]
```diff
@@ -0,0 +1,711 @@
...
+        (true, u128::from(n[0]) | (u128::from(n[1]) << 64))
+    }
+}
+
+/// GLV split: `k = k1 + k2·λ (mod r)` with |k1|, |k2| ≤ 2^128, each half
+/// returned as (is_negative, magnitude).
+fn decompose<C: GlvParams>(k: &C::ScalarExt) -> ((bool, u128), (bool, u128)) {
+    let repr = k.to_repr();
+    let bytes: &[u8] = repr.as_ref();
+    debug_assert_eq!(bytes.len(), 32, "Pasta scalars have a 32-byte repr");
```
Comment (ebfull) [inline src/glv.rs:200]: We know the representation is 32 bytes, because repr above is a `[u8; 32]` before `as_ref` right?

File: src/glv.rs [hunk 10]
```diff
@@ -0,0 +1,711 @@
...
+    let c2 = round_mul_shift(&C::G2, &kl);
+    // k1 = k − c1·V1A − c2·V2A   (two's complement over 256 bits)
+    let k1 = sub256(sub256(kl, mul_u128(c1, C::V1A)), mul_u128(c2, C::V2A));
+    // k2 = c1·V1B_NEG − c2·V2B   (v1.b = −V1B_NEG, v2.b = +V2B)
+    let k2 = sub256(mul_u128(c1, C::V1B_NEG), mul_u128(c2, C::V2B));
+    (signed_halves(k1), signed_halves(k2))
+}
+
+/// φ(P) on affine coordinates: (ζ·x, y). The identity maps to the identity.
+fn endo_affine<A: CurveAffine>(p: &A) -> A {
```
Comment (ebfull) [inline src/glv.rs:215]: this is a slow implementation of the endomorphism we already have implemented directly on the points: https://docs.rs/pasta_curves/latest/pasta_curves/arithmetic/trait.CurveExt.html#tymethod.endo

**Comment by nuttycom** [comment]
Superseded by #93
